components/ws-manager-mk2/controllers/workspace_controller_test.go

```diff
@@ -276,6 +277,87 @@ var _ = Describe("WorkspaceController", func() {
 			})
 		})
 
+		It("node disappearing should fail with backup failure", func() {
+			ws := newWorkspace(uuid.NewString(), "default")
+			m := collectMetricCounts(wsMetrics, ws)
+
+			// Simulate pod getting scheduled to a node.
+			var node corev1.Node
+			node.Name = uuid.NewString()
+			Expect(k8sClient.Create(ctx, &node)).To(Succeed())
+			// Manually create the workspace pod with the node name.
+			// We can't update the pod with the node name, as this operation
+			// is only allowed for the scheduler. So as a hack, we manually
+			// create the workspace's pod.
+			pod := &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       fmt.Sprintf("ws-%s", ws.Name),
+					Namespace:  ws.Namespace,
+					Finalizers: []string{workspacev1.GitpodFinalizerName},
+				},
+				Spec: corev1.PodSpec{
+					NodeName: node.Name,
+					Containers: []corev1.Container{{
+						Name:  "workspace",
+						Image: "someimage",
+					}},
+				},
+			}
+
+			Expect(k8sClient.Create(ctx, pod)).To(Succeed())
+			pod = createWorkspaceExpectPod(ws)
+			updateObjWithRetries(k8sClient, pod, false, func(pod *corev1.Pod) {
+				Expect(ctrl.SetControllerReference(ws, pod, k8sClient.Scheme())).To(Succeed())
+			})
+
+			markReady(ws)
+
+			// Make node disappear 🪄
+			By("deleting node")
+			Expect(k8sClient.Delete(ctx, &node)).To(Succeed())
+
+			// Expect workspace to disappear, with a backup failure.
+			// NOTE: Can't use expectWorkspaceCleanup() here, as the pod never disappears in envtest due to a nodeName being set.
+			// Therefore, we only verify deletion timestamps are set and all finalizers are removed, which in a real cluster
+			// would cause the pod and workspace to disappear.
```

Comment:
this took me a while to figure out 🤷 Only when `spec.nodeName` is set, the pod does not disappear from env test when it is requested to be deleted.